### DISCUSSION TITLE
[Testing] PS4 fpkg (*.pkg) support for HB-Store

### DISCUSSION THREAD

**Discussion by Gkiokan** [retronas/discussions#28] [Ideas]
Hi mate,

I just got a feedback that my [HB-Store CDN Server](https://github.com/Gkiokan/hb-store-cdn-server) tool would be great if it could supports RetroNAS as base path.  
I will test the RetroNAS once I have time for it to figure out how to implement it though.  
  
Currently there are a few endpoints and stuff to be made to get the pkgs working on the hb-store app on the ps4 side.  
So therefore my Tool would only act as a proxy here and prepare the necessary data structure and provide the needen stuff.
  
As my tool just deep scans the given base path for all *.pkg files, I assume that it would be enough if users just put in the NAS address as base path and let it do it's job.  Which seems to be working fine by the user.  But apperently it scans all the NAS folders I guess which adds a bit delay.

**Comment by sairuk** [comment]
Converting to discussion

Please let us know when this tool is ready to be added

**Reply by sairuk** [reply]
Added to [Ideas](https://github.com/danmons/retronas/wiki/Ideas)

**Comment by Gkiokan** [comment]
@sairuk HI mate. Thanks for the reply.  I almost forgot about that part.  
  
The tool is already finished and released and is working. 
You can find the latest release at [HB-Store CDN Server v1.7.2-RC](https://github.com/Gkiokan/hb-store-cdn-server/releases/tag/v1.7.2) 
Also I've created a CLI version [Hb-Store CDN Server CLI v1.2.0](https://github.com/Gkiokan/hb-store-cdn-cli-server/releases/tag/v1.2.0) 

Both work in the same way, so when we got it running on the one we can patch it on the other, too. 
 
Is there any changes to be done to load the files from the retroNAS itself? 
It would be enough if the retroNAS path would just return the *.fpkg files as I make a deep scan of the given path.  
 
If you need samples you can call me on discord `Gkiokan#0735`

**Comment by sairuk** [comment]
Can you publish build instructions for this application and an arm build? I see there is a linux-arm target in the git tree `package.json` but not in the last release version.

I have no experience with nodejs, nor the motivation to learn it at this time. 

I have attempted to build the git version with `pkg package.json -o test` but its been churning a single core at 100% and not doing much of anything that i can see for well over 20min which is too slow to build on device as part of install.

If i am doing something wrong please let me know.

**Reply by sairuk** [reply]
OK worked it out, should be able to automate it from here

```
$ git clone https://github.com/Gkiokan/hb-store-cdn-cli-server.git
$ cd hb-store-cdn-cli-server/
$ npm install #installs app deps
$ npm install pkg
$ npm run build:linux # or whatever target
$ cd release
$ ./hb-store-cdn-cli-server

------------------------------------------------------
  _   _ ____       ____  _                 
 | | | | __ )     / ___|| |_ ___  _ __ ___ 
 | |_| |  _ \ ____\___ \| __/ _ \| '__/ _ \
 |  _  | |_) |_____|__) | || (_) | | |  __/
 |_| |_|____/     |____/ \__\___/|_|  \___|
                                           
   ____ ____  _   _   ____                           
  / ___|  _ \| \ | | / ___|  ___ _ ____   _____ _ __ 
 | |   | | | |  \| | \___ \ / _ \ '__\ \ / / _ \ '__|
 | |___| |_| | |\  |  ___) |  __/ |   \ V /  __/ |   
  \____|____/|_| \_| |____/ \___|_|    \_/ \___|_|   
                                                     
------------------------------------------------------
 Version: 1.2.0
 HB-Store CDN Server [CLI version] Made by Gkiokan.NET
 GitHub: https://github.com/Gkiokan/hb-store-cdn-cli-server


```

**Reply by Gkiokan** [reply]
I would suggest to use the build release binaries directly.  
Building them require at least node v14.17.* and installing the depencendies first. 
I will apply the build instructions soon. Usually you would need to do 2 commands
`npm install`for installing the node depencendies.  
`npm run build`does build for the current system that you are on.  
You may also build for different OS / arch with  `npm run build:*`. A list of the build commands are below.
```
    "build": "npm run clean && npm run dist && pkg .",
    "build:all": "npm run clean && npm run dist && pkg . -t linux,macos,win",
    "build:win": "npm run clean && npm run dist && pkg . -t node14-win-x64",
    "build:mac": "npm run clean && npm run dist && pkg . -t node14-macos-x64",
    "build:linux": "npm run clean && npm run dist && pkg . -t node14-linux-x64",
    "build:linux-arm":
```

**Reply by sairuk** [reply]
thanks for the info

we can look at switching to the appropriate pre-built binaries if we run intoproblems. I couldn't do that here because our primary platform for RetroNAS is arm and there was no asset released for that build at 1.2.0 from what I could see support to build arm was added to packages.json after the 1.2.0 release.

the binary built fine without errors on debian 11 but this was with  npm 7.5.2 and nodejs 12.22 (assuming this equates to node12)
```
ii  nodejs                            12.22.12~dfsg-1~deb11u1          amd64        evented I/O for V8 javascript - runtime executable
ii  nodejs-doc                        12.22.12~dfsg-1~deb11u1          all          API documentation for Node.js, the javascript platform
ii  npm                               7.5.2+ds-2                       all          package manager for Node.js
```

you mention we need to build on node14 but i see no difference in the functionality for the minimal comparison i am able to do when built against what i assume as per above is node12

what problems can we expect to run into?

is there a way to shrink this app down from 170mb? that is a _lot_ of space for what it does, it blew the storage on my test RetroNAS install.

**Reply by Gkiokan** [reply]
I will also look into that. By the nature of node it does compile it's runtime into the final binary, too, so it is cross compatible to be run without dependencies. There are some compression settings that I can set and I will analyse the packages on build process to see what I can remove to shrink the size.   [Compression Task](https://github.com/Gkiokan/hb-store-cdn-cli-server/issues/6)

Once the build is done, you can remove the node_modules server as it takes also some space.   
 
Regarding node v12 <> v14 <> v17 there are some deprecations but luckly it does seem to work on v12, too. So it's fine then.

<img width="252" alt="image" src="https://user-images.githubusercontent.com/7249224/207540114-325d4d3c-ea83-4498-b430-9f1a1488e3f1.png">

**Comment by sairuk** [comment]
This has now been added to the `testing` branch and can be installed from the `Experimental` menu

This needs testing with someone who has an applicable setup, i can see the server running with pkg info listed

```
root@testronas:~# systemctl status hb-store-cdn.service 
● hb-store-cdn.service - Homebrew Store for PlayStation4
     Loaded: loaded (/lib/systemd/system/hb-store-cdn.service; enabled; vendor preset: enabled)
     Active: active (running) since Mon 2022-10-17 23:00:53 AEDT; 29min ago
   Main PID: 730176 (hb-store-cdn-cl)
      Tasks: 10 (limit: 4672)
     Memory: 43.5M
        CPU: 1.736s
     CGroup: /system.slice/hb-store-cdn.service
             └─730176 /opt/retronas/bin/hb-store-cdn/hb-store-cdn-cli-server start

Oct 17 23:00:54 testronas hb-store-cdn-cli-server[730176]: │ id         │ name         │ version │ size      │
Oct 17 23:00:54 testronas hb-store-cdn-cli-server[730176]: │ NPXS39041  │ Store        │ 01.00   │ 19.66 MB  │
Oct 17 23:00:54 testronas hb-store-cdn-cli-server[730176]: │ LAPY20001  │ PS4-Xplorer  │ 01.00   │ 73.53 MB  │
Oct 17 23:00:54 testronas hb-store-cdn-cli-server[730176]: └────────────┴──────────────┴─────────┴───────────┘
Oct 17 23:00:54 testronas hb-store-cdn-cli-server[730176]: ? ## Server ## HB-Store CDN CLI Server (Use arrow keys)
Oct 17 23:00:54 testronas hb-store-cdn-cli-server[730176]: ❯ CDN Server is: running
Oct 17 23:00:54 testronas hb-store-cdn-cli-server[730176]:   CDN Address: http://192.168.0.123:6449 
Oct 17 23:00:54 testronas hb-store-cdn-cli-server[730176]:   ──────────────
Oct 17 23:00:54 testronas hb-store-cdn-cli-server[730176]:   [Server] Start the server
Oct 17 23:00:54 testronas hb-store-cdn-cli-server[730176]:   [Server] Restart the server
root@testronas:~# 

```

**Reply by coldjim** [reply]
I've tested this quite a lot, and I can't get it to work. It just doesn't see the homebrew store running on my Unraid server (running Retronas as a VM)

**Reply by sairuk** [reply]
unfortunately my ps4 is on 9.6 so i can't HEN it still but here are some ideas to look at

does the service say its running?
```
$ systemctl status hb-store-cdn.service
     Active: active (running) since Sun 2022-12-11 07:20:42 AEDT; 17s ago

Dec 11 07:20:44 testronas hb-store-cdn-cli-server[52832]: ❯ CDN Server is: running
Dec 11 07:20:44 testronas hb-store-cdn-cli-server[52832]:   CDN Address: http://192.168.0.123:6449 

```

is it listening on 6449?
```
$ ss -lnp | grep 6449
tcp   LISTEN 0      511                                             *:6449                   *:*      users:(("hb-store-cdn-cl",pid=52832,fd=19))
```
if so the retronas side is OK


can you access the server manually with telnet from another machine?
```
$ telnet <retronas_ip> 6449
```
if not check ports are accessible through unraid/firewalls, i am not an unraid user so not sure if there is port stuff required or not.

if so then the retronas/unraid side as best we can offer without further detailed feedback on testing

**Reply by coldjim** [reply]
Hey,
Thanks for getting back to me quickly.

Yes the service is running fine shown here:
<img width="1073" alt="Screenshot 2022-12-10 at 22 33 27" src="https://user-images.githubusercontent.com/104848919/206878339-6d6e95cd-8f5a-41be-a5d5-f94ddf8298ef.png">

I have ran that command and found that it is also listening on 6449:
<img width="1012" alt="Screenshot 2022-12-10 at 22 32 06" src="https://user-images.githubusercontent.com/104848919/206878356-d8b22299-a07b-4cc1-9bc7-fb70d0f4f889.png">

Yes I can also access it via telnet from my macbook terminal (after installing telnet through homebrew). Covered my macbook name but it can be seen here, tried 6551 as a test also to make sure telnet works and it refuses to connect:
<img width="429" alt="Screenshot 2022-12-10 at 22 50 08" src="https://user-images.githubusercontent.com/104848919/206878386-3bb43c4a-7bdd-432c-908b-b18f552a2c39.png">

I am willing to keep testing to hopefully make this work. Would be a great addition to have, since the aim is to have everything in one place.
The only thing I had to change was the config.ini within the HBstore to look for my basepath (as its not /data).
<img width="415" alt="Screenshot 2022-12-10 at 22 44 59" src="https://user-images.githubusercontent.com/104848919/206878585-56a15c5d-ba32-4f0d-8b80-007cec5965a5.png">

I have also attached the logs from the HB store that can be pulled from the PS4 via FTP. Maybe @Gkiokan can give more insight into this? My hunch is that the PS4 can't find the 'store.db'? Unsure why at this point.

[log.txt](https://github.com/danmons/retronas/files/10201125/log.txt)
[loader.log](https://github.com/danmons/retronas/files/10201126/loader.log)

Anything else you would like me to test, let me know.
The ports on this VM have had no issues so far, as I installed Samba and Cockpit using Retronas so far and these both work fine.

**Reply by sairuk** [reply]
thanks the path in the config file was an oversight on my behalf, i've pushed a fix for that to testing branch now. appreciate you taking the time to test

**Reply by sairuk** [reply]
based on this from the log

```
[STORE_GL_Loader:main:453] ----- CDN Url = http://192.168.4.99 ---
[STORE_GL_Loader:pingtest:124] ----- sceHttpsSetSslCallback() Success ---
[STORE_GL_Loader:pingtest:133] ----- sceHttpCreateConnectionWithURL() Success ---
[STORE_GL_Loader:main:460] ----- Ping Successfully ---
connection errror
[STORE_GL_Loader:download_file:303] ----- status code = 404 ---
The Store Loader has encountered an error
```


seems to be a similar issue but not info https://github.com/LightningMods/PS4-Store/issues/15 since the user didn't do anything id assume it was a backend CDN issue. in this have the hb-store-cdn app may need a change

**Reply by Gkiokan** [reply]
Hellow guys, thanks for testing.

Seems like the hb store did connect to the server but can't get any data out of it?

Could you delete the log on the ps4 and make a new test and share the full log?

I can have a look into that tomorrow and check if the native way would work and then compare what's missing within the retroNAS config.

FYI as a general information.
The ps4 hb store does use a store.db file that will be checked and updated on each start of the hb store on the ps4. This file and the target paths will be provided via the cdn tool on the selected port. Where as the paths for the pkg files would be the given base path + filename from the cdn. 

Could you try to download the store.db (sqlite DB file) from the cdn Address with http://your_cdn:port/store.db and check the paths?

**Reply by sairuk** [reply]
i checked this store.db endpoint earlier after looking through the hbstore source. on a fresh install atop RetroNAS the store.db is served to the browser fine. been a bit busy so forgot to get back to this topic to update it.

The store.db file also looked to be correct when queried with sqlite3 as the package data was listed in the homebrews table.

grabbed multiple times
```
$ file Downloads/store*
Downloads/store(1).db: SQLite 3.x database, last written using SQLite version 3038002
Downloads/store(2).db: SQLite 3.x database, last written using SQLite version 3034001
Downloads/store.db:    SQLite 3.x database, last written using SQLite version 3038002
```

```
$ curl -IL http://testronas:6449/store.db
HTTP/1.1 200 OK
X-Powered-By: Express
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, PATCH, DELETE
Content-Disposition: attachment; filename="store.db"
Accept-Ranges: bytes
Cache-Control: public, max-age=0
Last-Modified: Sun, 11 Dec 2022 04:08:48 GMT
ETag: W/"10000-184ff5f05bf"
Content-Type: application/octet-stream
Content-Length: 65536
Date: Sun, 11 Dec 2022 04:14:06 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```

sorry should also add i can pull down pkgs as well with the browser/curl from the RetroNAS install

```
$ curl -IL http://testronas:6449/IV0001-PFBN00001_00-PFBN000016050000.pkg
HTTP/1.1 200 OK
X-Powered-By: Express
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, PATCH, DELETE
Content-Disposition: attachment; filename="IV0001-PFBN00001_00-PFBN000016050000.pkg"
Accept-Ranges: bytes
Cache-Control: public, max-age=0
Last-Modified: Tue, 30 Aug 2022 13:16:48 GMT
ETag: W/"75c0000-182eee5b600"
Content-Type: application/octet-stream
Content-Length: 123469824
Date: Sun, 11 Dec 2022 04:36:03 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```

we are also sending the data OK

server
```
05d35d974fac3c65d2318b976ed5fc00  IV0001-PFBN00001_00-PFBN000016050000.pkg
```

client
```
05d35d974fac3c65d2318b976ed5fc00  IV0001-PFBN00001_00-PFBN000016050000.pkg
```

**Reply by sairuk** [reply]
Does anyone know what headers etc the PS4 sends so I can attempt to simulate the failure?

```
User-Agent: StoreHAX/GL-0x9008000
```

**Reply by sairuk** [reply]
comparing the sql between CDNs shows results are OK and working. the package entry from this tool successfully downloads the file from the RetroNAS

hb-store-cdn
```
sqlite> select id,package from homebrews where id = 'PFBN00001';
PFBN00001|http://192.168.0.123:6449/IV0001-PFBN00001_00-PFBN000016050000.pkg
sqlite> 
```

pkg-zone (retrieved with `curl -L -O -H'User-Agent: StoreHAX/GL-0x9008000'  http://api.pkg-zone.com/store.db`)
```
sqlite> select id,package from homebrews where id = 'PFBN00001';
PFBN00001|http://api.pkg-zone.com/detail/PFBN00001
sqlite> 
```

**Reply by coldjim** [reply]
Ok, i might not be able to do any testing today but is there anything you'd like me to try when I next get some time? I will delete the logs and pull them for you @Gkiokan

**Reply by coldjim** [reply]
When i type '192.16.8.4.30:6449/store.db' into my browser, it downloads the store.db file fine. It must be the ps4 end that isn't seeing this. Unsure how that bit works

**Reply by Gkiokan** [reply]
So the CDN does in fact works correctly. You did a great job on digging into that. 
The StoreHAX header is only used in the public api, but does have no meaning on the CDN. 
Yea we must see why the ps4 have issues with that. Luckly @LightningMods does a full log of this HB-Store app so we can see whats happening. 

We would need a clean log file from your ps4 of the hb-store logs. 
Also do you have any other configuration set besides the CDN address for the hb-store app on the config.ini (ps4)?

**Reply by sairuk** [reply]
> When i type '192.16.8.4.30:6449/store.db' into my browser, it downloads the store.db file fine. It must be the ps4 end that isn't seeing this. Unsure how that bit works

your previous retronas ip address was `192.168.4.99` as per your ps4 config and it is now `.30`? can you please confirm.

**Reply by coldjim** [reply]
> > When i type '192.16.8.4.30:6449/store.db' into my browser, it downloads the store.db file fine. It must be the ps4 end that isn't seeing this. Unsure how that bit works
> 
> your previous retronas ip address was `192.168.4.99` as per your ps4 config and it is now `.30`? can you please confirm.

Sorry this was a typo on my part. I meant to put .99 rather than .30 (.30 is the unraid server, .99 is the retronas VM). I've just tested again to confirm and it works downloading the store typing "http://192.168.4.99:6449/store.db". I've got a little bit of time now, so i will pull the logs from the PS4 again and do some testing

**Comment by LightningMods** [comment]
Hello, if you are having issues be sure the ini has the keyword `http://` like `http://192.16.8.4.30:6449/store.db` if it still doesn't work you can send me the log from your PS4 using FTP so I can find the issue for you, PS4 Path:  `/user/app/NPXS39041/logs/log.txt`

**Reply by coldjim** [reply]
You mean the config within the homebrew store app? Or within the config.ini on the server? I changed it on the server as a test and still no luck

**Reply by LightningMods** [reply]
can send me the log from your PS4 using FTP so I can find the issue for you, PS4 Path: `/user/app/NPXS39041/logs/log.txt`, the log is auto-generated by the app

**Reply by Gkiokan** [reply]
The config on your ps4.  Should be in `/user/app/NPXS39041` when I remember right.  Paste the config and logs here so we can analyse it.

**Reply by coldjim** [reply]
> can send me the log from your PS4 using FTP so I can find the issue for you, PS4 Path: `/user/app/NPXS39041/logs/log.txt`, the log is auto-generated by the app

[loader.log](https://github.com/danmons/retronas/files/10221080/loader.log)
[log.txt](https://github.com/danmons/retronas/files/10221081/log.txt)

These are fresh logs from a new install with 192.168.4.99 in my config file

**Reply by coldjim** [reply]
[settings.ini.zip](https://github.com/danmons/retronas/files/10221099/settings.ini.zip)

Settings.ini from my PS4 also (had to zip it)

**Reply by coldjim** [reply]
@LightningMods @Gkiokan 
Weirdly... i updated my settings.ini on my PS4 to "http://192.16.8.4.30:6449/store.db" as suggested and now my server seems to see the PS4 IP as I get this:
<img width="1148" alt="Screenshot 2022-12-13 at 18 45 32" src="https://user-images.githubusercontent.com/104848919/207419069-479622d5-043a-4301-bbeb-9cf6dbe4da7a.png">

But i still get an error

**Reply by sairuk** [reply]
what happens if you enter `http://192.16.8.4.30:6449` (no store.db) in your ps4 settings file?

that output is triggered when _a_ device (not necessarily a ps4) accesses the `store.db` endpoint

i don't know why i didn't think of that the other day, the server runs on a custom port where as api.pkg-zone will be listening on standard ports. so you'd need the ip and port in your settings ini. We do mention the custom port in the doco but based on this discussion we can make that clearer.

```
[Settings]
CDN=http://192.168.4.99:6449
```

404 error return threw me i think

```
2022-12-13 18:49:35 INFO  source/utils.c:507: set->opt[CDN_URL ]: http://192.168.4.99
```
should read
```
INFO  source/utils.c:507: set->opt[CDN_URL ]: http://192.168.4.99:6449
```

**Reply by sairuk** [reply]
i'm going to guess you have one of the lightttpd based services installed like 3DS qr codes that listen on port 80 and that is what was returning the 404 _not_ the hb-store-cdn tool.

**Reply by sairuk** [reply]
i've added some updates to our doco with info from this thread on how to troubleshoot issues and a modded copy of the `settings.ini` you attached @coldjim thanks. hopefully its a bit clearer now for others

https://github.com/danmons/retronas/wiki/hb-store-cdn

**Reply by coldjim** [reply]
OK thats great, but it still doesn't work for me. It still throws up an error on the PS4, so something still isn't right. 
The server can see the PS4 now after editing the settings.ini file, but it still says 'The Store Loader has encountered an error'

And yes I have the lightttpd services installed.

**Reply by sairuk** [reply]
> 'The Store Loader has encountered an error'

~is this an error on the PS4 side?~ nvm you mentioned it was on the PS4 side already.

Please regenerate your package list after updating the settings from the RetroNAS menu and try again.

Can you also post new logs if not fixed for everyone to review?

**Reply by coldjim** [reply]
I have a new error now after a bit of messing around. 
<img width="1129" alt="Screenshot 2022-12-13 at 20 28 37" src="https://user-images.githubusercontent.com/104848919/207437083-7f4a01ea-00a0-4433-99c2-8f84f00b1f49.png">

**Reply by coldjim** [reply]
Is there supposed to be a remote.md5 file?

**Reply by sairuk** [reply]
ok that is something to work with, perhaps @Gkiokan can let me know what that file is for and how it is generated.

Looks like it is supposed to be pulled from the main project releases page looking at the source [here](https://github.com/Gkiokan/hb-store-cdn-server/blob/58a8d61e15bc4513e3dad9580438c99512a31428/src/plugins/hb.js#L11)

**Reply by sairuk** [reply]
looks like i need to force download these bins at startup and then start the server. let me look into that

**Reply by coldjim** [reply]
@sairuk 
![20221213_203449](https://user-images.githubusercontent.com/104848919/207438298-1e513589-af64-413a-bad4-bf90e78707cc.jpg)

SUCCESS!!
I had to get the windows version of the 'hb-store-cdn-cli-server' (couldn't get the mac one to work), and one of the options is to download the latest 'bin' files. I put them in the bin folder on my retronas folder that the server was looking in and it worked

**Reply by sairuk** [reply]
@Gkiokan doesn't look like i can force a bin dl from the cli, 

I think the app should be checking if they exist and pulling them if not and also have the option to update the bins from the cli available as well. that way i can do something like this for the service

```
# /lib/systemd/system/hb-store-cdn.service
[Unit]
Description=Homebrew Store for PlayStation4
After=network.target
StartLimitIntervalSec=60
StartLimitBurst=4

[Service]
User=retronas
WorkingDirectory=/opt/retronas/bin/hb-store-cdn
ExecStartPre=-/opt/retronas/bin/hb-store-cdn/hb-store-cdn-cli-server download-bin
ExecStart=/opt/retronas/bin/hb-store-cdn/hb-store-cdn-cli-server start
Restart=on-failure
RestartSec=1
SuccessExitStatus=3 4
RestartForceExitStatus=3 4

[Install]
WantedBy=multi-user.target
```

**Reply by coldjim** [reply]
For future reference:

The 'settings.ini' on my PS4 had to look at: http://192.168.4.99:6449
The 'config.ini' on the retronas server is unchanged:
host=192.168.4.99
port=6449
basePath=/mnt/retronas/ps4/hb-store-cdn
binVersion=0.00

Then the bin files are needed. There are 3 files for this. I can attach them here if you like?

**Reply by sairuk** [reply]
> SUCCESS!! I had to get the windows version of the 'hb-store-cdn-cli-server' (couldn't get the mac one to work), and one of the options is to download the latest 'bin' files. I put them in the bin folder on my retronas folder that the server was looking in and it worked

it's been a journey, great work sorting it out. there is some missing functionality to run this as a service as it turns out (the bin downloads). hopefully @Gkiokan can help us get that added

**Reply by sairuk** [reply]
> Then the bin files are needed. There are 3 files for this. I can attach them here if you like?

that is fine, thanks, no need. I can see the missing functionality and I could patch in support but I think I'll give @Gkiokan a chance to catch up to our latest discoveries.

The RetroNAS _could_ download this if the server is run interactively

**Reply by coldjim** [reply]
haha it really has been a journey. Thanks for your help on it, really loving retronas and its something i've wanted for a while, to have everything in one place.
Ye once you have those bin downloads sorted, i think you should be good. Now to fill the server with some games :)

**Reply by Gkiokan** [reply]
Ouh right, there was that binaries thing that you need to call first so the server can do it's validity checks. And yes the remote files etc. are needen. I totally forgot about that part.  
If you run the CLI tool on your commandline/terminal you have this options to download binaries, it should been also mentioned somewhere in the readme, it has been a while since I finished the project.

So the right move is
1) trigger binary dowload
2) start the server
-> this gives you as output the URI of the CDN as http://yourip:port -> this needs to be added into the settings.ini in your ps4 for the hb-store.

Once this is setup, everything should work. I will setup a better guide to start this or can you add that flow as a bash file when the service is starting up on the retroNas @sairuk?

**Reply by Gkiokan** [reply]
<img width="426" alt="image" src="https://user-images.githubusercontent.com/7249224/207441740-db8cf762-473b-4c98-8189-b9526cd32568.png">

If you start the tool from your terminal you can select this option `Check Server Binaries` or `Force re-download server binaries` this is actually necessary, sorry about that. Missed it totally, thought it was clear.  I will look up later if I have added a direct way to call the binary download method directly from a command, so you can put that in a starting script @sairuk.

Btw, it's a express server that runs on the given port to handle the file streams.

**Reply by sairuk** [reply]
we will need a non-interactive version of this functionality so we can trigger the `download-bin` operation. The user never sees the menu when its run as a service so much like the `start` functionality on the cli i would be looking for a `download-bin` option to trigger the functionality from the cli prior to calling `start`

**Reply by sairuk** [reply]
can we add a `download-bin` connection in here for example? 

https://github.com/Gkiokan/hb-store-cdn-cli-server/blob/0561f82769fdc86ba215263855cd6c37fb63e81f/src/cli.js#L207

**Reply by Gkiokan** [reply]
Maybe I can combine that both into one command that does both at once.

But we must take one thing into account. Once the HB-Store gets an update, the binaries that needs to be downloaded again as they change. I guess the service is running once for ever when it was started? Is there a way to restart the service for the user within retroNAS? For the use-case once the store has an update, which aren't that often. 

I will get my hands on this maybe in 1-2 days and commit a new version.

**Reply by coldjim** [reply]
@Gkiokan just a question about how the HB store works...at the moment you have to download the PKG to either the PS4 internal hard drive or a USB drive (if you change the setting to True). Is there anyway to skip the download bit and just install straight from the server if you put the path to it? Or is this not possible?

**Reply by sairuk** [reply]
>  Is there a way to restart the service for the user within retroNAS? 

yeh the service will run so long as RetroNAS is up.

We have interactive start/stop functionality for our services in the RetroNAS menu and refreshing the package list also triggers a restart. 

If you were to expose the check update functionality as well with appropriate exit codes we could run that on a timer which would trigger a service reload based on the result.

my suggestion for available cli args based on this discussion are
```
start -> start the server
stop -> stop the server
status -> check server state

download-bin -> force download of required binaries
check -> check for binary updates
```

keep them separate and we can call them as necessary. 

We could have an `ExecStartPre` action to download/update the required files on service start or a wrapper script to run check before download-bin, we could also set up a timer to check for updates.

**Reply by Gkiokan** [reply]
@coldjim This are 2 different things. 

The settings.ini is on the PS4 Storage itself, where the app is installed, no need to use a USB here.   
That configuration has to be made once to tell the HB-Store where it should get the source data from.   

The second part is the HB-Store (CLI) Server, which needs this binary files on the server side (your machine or in this case retroNAS). It is crucial as the HB-Store App on the ps4 does check this files for validity and stuff. LM can tell us more about that, but without this it doesn't work.  

I have just tried to merge both methods into one command but that doesn't work as expected. I need to make a Promise wrapper between them and call them sequentially to get run in the correct order without exiting the code. Another way would be a double call like ./app bin && ./app start but that isn't an optimal solution.  

So the best solution is that I will integrate another command that does check the binary files infront before it starts the server.  
I will do both and maybe add the binary check on the startup. Usually users know when they have to update it as it is a rare to call method.

**Reply by Gkiokan** [reply]
@sairuk yea sounds good to me. I will provide you all methods to be callable with args and also add another fully automated method  to start it with the check in front. Gimme 1-2 days I will implement that and push a new conversation then.   

You do create the config.ini so far correctly when the service is installed? Like the host, port and basePath ?  

For another feature we could also have a method that downloads the log file from the PS4 but we need the PS4 IP to bet set here or I can grab it once the ps4 does connect to the service. I did something similar on the [GUI version.](https://github.com/Gkiokan/hb-store-cdn-server)

**Reply by sairuk** [reply]
> @sairuk yea sounds good to me. I will provide you all methods to be callable with args and also add another fully automated method to start it with the check in front. Gimme 1-2 days I will implement that and push a new conversation then.

sounds good, thank you.

> You do create the config.ini so far correctly when the service is installed? Like the host, port and basePath ?

our config file is fine based it off what was created at server launch, you can see our defaults here. the "{{ xxxx }}" are jinja vars that are populated at install time based on data available through ansible facts

https://github.com/danmons/retronas/blob/main/ansible/templates/install_hb-store-cdn/config.ini.j2

we set bin version to `0.00`. in theory i wanted that to trigger an update but will admit i didn't fully understand the requirement based on the findings here. 

> For another feature we could also have a method that downloads the log file from the PS4 but we need the PS4 IP to bet set here or I can grab it once the ps4 does connect to the service. I did something similar on the [GUI version.](https://github.com/Gkiokan/hb-store-cdn-server)

I would go with the latter. if you do provide this functionality and expose it to the cli we can look into adding an option to trigger it and upload it to their RetroNAS which they could then retrieve off of their samba shared path if they didn't want to use ftp.

**Reply by Gkiokan** [reply]
The initial config with the params that you provide are fine.  
Also the version 0.00 is okay as I make a version compare from the config version against the remote version.  

Good job guys, good work. See ya in the next update in a couple days, have a nice day.

**Reply by sairuk** [reply]
i've just hacked in the force download option and the check option for cli which was pretty straightforward. a small function in `cli.js` and a conditional call in `app.js`. i am not particularly familiar with node so just hacked something in. The additional `check`/`check-and-download` options are updates to `bin.js` to return `compare` and using that as a condition and an exit code

download
```
[Bin] Force re-download Server Binaries. Update in progress
[Bin] Found 3 to download. Downloading ...
[Main] Saved config to /opt/retronas/bin/hb-store-cdn/config.ini
[Bin] Download finished.
```
check
```
------------------------------------------------------
[Bin] Compare Current Version: 4.30 <> GitHub version: 4.31
[Bin] New Server Binaries are available. Update is required
root@testronas:/opt/retronas/src/hb-store-cdn# echo $?
255

[Bin] Compare Current Version: 4.31 <> GitHub version: 4.31
[Bin] Your Server Binaries are up to date
root@testronas:/opt/retronas/src/hb-store-cdn# echo $?
0
```

check-and-download
```
[Bin] Compare Current Version: 4.31 <> GitHub version: 4.31
[Bin] Your Server Binaries are up to date
[Main] No download required

[Bin] Compare Current Version: 4.30 <> GitHub version: 4.31
[Bin] New Server Binaries are available. Update is required
[Bin] Force re-download Server Binaries. Update in progress
[Bin] Found 3 to download. Downloading ...
[Main] Saved config to /opt/retronas/bin/hb-store-cdn/config.ini
[Bin] Download finished.
```

i did notice that you are not checking the files actually exist on the filesystem but only seem to be relying on the ini version and remote version check. So if the files are removed from the filesystem the ini needs to be corrected or a 'force' run

I'd probably have the check as exactly what it says, check the version and just notify the user of update availablity don't also download them, have that check and update option you mentioned in the menu

**Reply by coldjim** [reply]
Is there anyway to not have to 'manually' refresh the package folder every time a PKG is added to the basefolder? Can it auto refresh every so often?

**Reply by sairuk** [reply]
> Is there anyway to not have to 'manually' refresh the package folder every time a PKG is added to the basefolder? Can it auto refresh every so often?

good question, 

yes RetroNAS could use a systemd path unit to monitor for changes although that depends on if the hb store folder layout supports a folder hierarchy or not, if it does we wont be adding it because we need a path unit per sub directory which is a limitation of systemd annoyingly

if however it is just a single depth folder for where the pkg sits we could do something for auto refresh

i assume however the pkg list could be reloaded internally by the server if the modified dates change as well, not sure how heavy that would be.

**Reply by Gkiokan** [reply]
The HB-Store CLI Server does read all files below the basePath. The responsible code part is here `let files = fg.sync([patchedBasePath + '/**/*.pkg'])`

As for the nature of the PS4 pkgs beeing usually pretty big, mostly 4GB+ up to 50GB and more, you don't modify them that often.   
Simple fix would be just to restart the server here to trigger a rescan.  
 
But I can add a new feature in the future to the application itself to watch for file changes and apply the updates to the current state of the server, so any changes will be reflected accordingly. But you need to sync your HB-Store on your PS4 then, too. Because it uses the store.db as the database to lookup for the existing games on the remote server. As far as I know this has been changed from api calls to store.db to improve performance in general.  Because of that the "refersh" or "watch" feature was never touched really from me cuz if you need to change something, you usually did a restart on both, server and hb-store on the ps4.  

[Task to watch basePath](https://github.com/Gkiokan/hb-store-cdn-cli-server/issues/8)  
[Task to expose commands](https://github.com/Gkiokan/hb-store-cdn-cli-server/issues/7)

**Reply by Gkiokan** [reply]
@sairuk Have a look into the new [release v1.3.0](https://github.com/Gkiokan/hb-store-cdn-cli-server/releases/tag/v1.3.0)

I've managed to add the binary check to the start parameter so it will be checked always before start.  
But you can also check the exposed params that allows you to handle the underlying methods for e.g. trigger binary download.   
I wasn't sure if I should send a exit code in specific cases, but the logs itself does contain either Download Finished or You are on the latest version if this may help already?  

But anyways as the code does now does check the binary on startup I would recommend only to restart the service for updates instead of doing an interval check.  The watch method isn't implemented yet, but this could be done on my side within the app that it will a) watch for changes in the basePath folder and b) updates on the binary files and does the update itself without extra triggering.

**Reply by sairuk** [reply]
that looks like a good update, i will take a look in the next day or so

exit codes are helpful to determine the result of commands without the additional overhead of querying logs, but i will take a look how it works now and see.

**Reply by sairuk** [reply]
i've updated our implementation with a check script that runs every 12h and will download/restart if required

**Reply by coldjim** [reply]
@sairuk great work, thanks. I will update to this soon. Is this in the testing branch?

**Reply by sairuk** [reply]
yeh testing branch, let me know how you go.
